blockchain/stake/staketx.go

```diff
@@ -707,10 +710,128 @@ func IsSStx(tx *wire.MsgTx) bool {
 	return CheckSStx(tx) == nil
 }
 
-// CheckSSGen returns an error if a transaction is not a stake submission
+// TreasuryVoteT is the type that designates a treasury vote. There are two
+// valid bits that may be set, although not simultaneously. 0x01 and 0x02. Any
+// other (or lack therefore) bits are considered invalid.
+type TreasuryVoteT byte
+
+const (
+	TreasuryVoteInvalid TreasuryVoteT = 0x00 // Invalid vote
+	TreasuryVoteYes     TreasuryVoteT = 0x01 // Vote YES
+	TreasuryVoteNo      TreasuryVoteT = 0x02 // Vote NO
+)
+
+// CheckTreasuryVote ensures that the provided treasury vote is valid. If the
+// vote is valid the proper type is returned.
+func CheckTreasuryVote(vote TreasuryVoteT) (TreasuryVoteT, error) {
+	switch vote {
+	case TreasuryVoteYes:
+	case TreasuryVoteNo:
+	default:
+		return TreasuryVoteInvalid,
+			fmt.Errorf("invalid treasury vote: 0x%x", vote)
+	}
+	return vote, nil
+}
+
+// IsTreasuryVote returns true if the provided vote is valid.
+func IsTreasuryVote(vote TreasuryVoteT) bool {
+	_, err := CheckTreasuryVote(vote)
+	return err == nil
+}
+
+// TreasuryVoteTuple is a tuple that groups a TSpend hash and its associated
+// vote bits.
+type TreasuryVoteTuple struct {
+	Hash chainhash.Hash
+	Vote TreasuryVoteT
+}
+
+// GetSSGenTreasuryVotes pulls out treasury votes for TSpend transactions. This
+// function is a convenience function to discourage rolling new versions that
+// pull out votes and may be therefore different from consensus.
+// This function verifies that the passed in script is of the following form:
+// OP_RETURN DATA_PUSH []byte{'T', 'V'} N-votes
+// and returns a slice of hashes which represent the TSpend transactions they
+// are supposed to vote on.
+func GetSSGenTreasuryVotes(PkScript []byte) ([]TreasuryVoteTuple, error) {
+	// Verify that there is enough length to contain a discriminator.
+	// Expect at least: OP_RETURN DATA_PUSH Y Z
+	if len(PkScript) < 4 ||
+		PkScript[1] > txscript.OP_PUSHDATA1 {
+		str := fmt.Sprintf("last SSGen cannot not contain a " +
+			"type discriminator")
+		return nil, stakeRuleError(ErrSSGenInvalidDiscriminatorLength,
+			str)
+	}
+
+	// Determine start of discriminator based on the opcode in [1].
+	start := 2
+	if PkScript[1] == txscript.OP_PUSHDATA1 {
+		start = 3
+	}
+
+	// Ensure discriminator is TV.
+	if !bytes.Equal(PkScript[start:start+2], []byte{'T', 'V'}) {
+		str := fmt.Sprintf("last SSGen unknown type discriminator: "+
+			"0x%x 0x%x", PkScript[start], PkScript[start+1])
+		return nil, stakeRuleError(ErrSSGenUnknownDiscriminator, str)
+	}
+
+	// Since this is a 'T','V' we expect N hashes and their vote bits.
+	const size = chainhash.HashSize + 1
+	if len(PkScript[start+2:]) < size ||
+		len(PkScript[start+2:])%size != 0 {
+		str := fmt.Sprintf("SSGen 'T','V' invalid " +
+			"length")
+		return nil, stakeRuleError(ErrSSGenInvalidTVLength,
+			str)
+	}
+
+	// Return hashes, this is the success path.
```

Comment:
No need to check that the DATA_PUSH had the correct value?

Reply:
Fixed.